core/include/scipp/core/values_and_variances_strong_types.h

```diff
@@ -0,0 +1,202 @@
+  Variances(Args &&... args)
+      : VariancesTaggedTupple<Args...>(
+            detail::makeTaggedTuple<tag_type>(std::forward<Args>(args)...)) {}
+  template <class T>
+  Variances(std::initializer_list<T> init)
+      : VariancesTaggedTupple<Args...>(
+            detail::makeTaggedTuple<tag_type>(init.begin(), init.end())) {}
+};
+template <class... Args> Variances(Args &&... args)->Variances<Args...>;
+template <class T>
+Variances(std::initializer_list<T>)
+    ->Variances<typename std::initializer_list<T>::iterator,
+                typename std::initializer_list<T>::iterator>;
+
+namespace detail {
+template <class Tag, class T> struct has_tag : std::false_type {};
+
+template <class Tag, class... Ts>
+struct has_tag<Tag, Values<Ts...>> : std::is_same<Tag, ValuesTag> {};
+
+template <class Tag, class... Ts>
+struct has_tag<Tag, Variances<Ts...>> : std::is_same<Tag, VariancesTag> {};
+
+template <class T, class... Args>
+constexpr bool is_type_in_pack_v =
+    std::disjunction<std::is_same<T, std::decay_t<Args>>...>::value;
+
+template <class Tag, class... Args>
+constexpr bool is_tag_in_pack_v =
+    std::disjunction<has_tag<Tag, Args>...>::value;
+
+template <class T, template <class T1, class T2> class Cond, class... Args>
+class Indexer {
+  template <std::size_t... IS>
+  static constexpr auto indexOfCorresponding_impl(std::index_sequence<IS...>) {
+    return ((Cond<T, Args>::value * IS) + ...);
+  }
+
+public:
+  static constexpr auto indexOfCorresponding() {
+    return indexOfCorresponding_impl(
+        std::make_index_sequence<sizeof...(Args)>{});
+  }
+};
+
+template <class T, class Tuple, std::size_t... I>
+constexpr T make_move_from_tuple_impl(Tuple &&t, std::index_sequence<I...>) {
+  return T(std::move(std::get<I>(t))...);
+}
+
+template <class T, class Tuple> constexpr T make_move_from_tuple(Tuple &&t) {
+  return detail::make_move_from_tuple_impl<T>(
+      std::forward<Tuple>(t),
+      std::make_index_sequence<
+          std::tuple_size_v<std::remove_reference_t<Tuple>>>{});
+}
+
+template <class VarT, class... Ts> class ConstructorArgumentsMatcher {
+public:
+  template <class... NonDataTypes> constexpr static void checkArgTypesValid() {
+    constexpr int nonDataTypesCount =
+        (is_type_in_pack_v<NonDataTypes, Ts...> + ...);
+    constexpr bool hasVal = is_tag_in_pack_v<ValuesTag, Ts...>;
+    constexpr bool hasVar = is_tag_in_pack_v<VariancesTag, Ts...>;
+    static_assert(
+        nonDataTypesCount + hasVal + hasVar == sizeof...(Ts),
+        "Arguments: units::Unit, Shape, Dims, Values and Variences could only "
+        "be used. Example: Variable(dtype<float>, units::Unit(units::kg), "
+        "Shape{1, 2}, Dims{Dim::X, Dim::Y}, Values({3, 4}))");
+  }
+
+  template <class ElemT, class... NonDataTypes>
+  static VarT construct(Ts &&... ts) {
+    auto tp = std::make_tuple(std::forward<Ts>(ts)...);
+    auto val = std::move(extractTagged<ValuesTag, Ts...>(tp).tuple);
```

Comment:
You know `ElemT` here, so you could directly create the correct `val` and `var`? This would remove a lot of templating from `createVariable`, for better compile times and code size. You would also need to move the `if constexpr (std::is_constructible_v<Vector<T>, T1 &&...> /*...*/` here.

Reply:
It would work, but I don't want to bring any knowledge about `Vector` on this file.